MAINT: fix domain edge error for betainc a or b ==0 case is 1.

rlucas7 opened the pull request:
#### Reference issue
Closes gh-8411

#### What does this implement/fix?
the `betainc` function was not providing correct values at boundary values for `a` and `b`, when either are `0.0`. In the described boundary case the values returned should be `1.0`. This PR implements the change. Also adds a test module to confirm the boundary values are as expected. 


#### Additional information
Previously (taken from gh-8411):
```
import numpy as np
from scipy.special import betainc

X = 0.5
Z = np.array(range(0,11))
W = 3

print(betainc(Z,W,X))
[       nan 0.875      0.6875     0.5        0.34375    0.2265625
 0.14453125 0.08984375 0.0546875  0.03271484 0.01928711]
```

Now: 
```
(scipy-dev) Lucass-MacBook:scipy rlucas$ python 
Python 3.6.7 |Anaconda, Inc.| (default, Oct 23 2018, 14:01:38) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> from scipy.special import betainc
>>> X = 0.5
>>> Z = np.array(range(0,11))
>>> W = 3
>>> print(betainc(Z,W,X))
[1.         0.875      0.6875     0.5        0.34375    0.2265625
 0.14453125 0.08984375 0.0546875  0.03271484 0.01928711]
>>> 
(scipy-dev) Lucass-MacBook:scipy rlucas$
```

person142 commented on scipy/special/cephes/incbet.c:
```diff
@@ -72,15 +72,17 @@ static double incbcf(double a, double b, double x);
+    if (aa < 0.0 || bb < 0.0)
 	goto domerr;
 
+    if (aa == 0.0 || bb == 0.0)
```
This can't be quite right-now you're going to get

```python
sc.betainc(1e-300, 1e-300, 0) == 0
```

but

```python
sc.betainc(0, 0, 0) == 1
```

i.e. there's a discontinuity, which should result in NaN.

I think we're also going to need some justification that this does the right thing for `a == 0` but `b == inf` say. Generally speaking there's a cube

```
a         b         x
[0, oo] x [0, oo] x [0, 1]
```

and each of the 6 faces, 12 edges, and 8 vertices needs to be thought about.

rlucas7 replied:
@person142  Thanks! Your right, I'd been thinking this was just a fix for an isolated point but will require some more thinking, and maybe I need to look at a contour integral or two to understand the behavior along the faces of this space.

WarrenWeckesser replied:
There might be a consistent way to define the values for `a == 0, b > 0` or for `a > 0, b == 0`, but for `a == 0, b == 0`, the value is not well-defined.  You can see this numerically by computing the function for a series of parameter values (a(t), b(t)) that approach (0, 0) along different straight lines:

```
In [140]: t                                                                               
Out[140]: array([1.e-06, 1.e-07, 1.e-08, 1.e-09, 1.e-10, 1.e-11, 1.e-12, 1.e-13])

In [141]: betainc(t, t, 0.3)
Out[141]: 
array([0.49999958, 0.49999996, 0.5       , 0.5       , 0.5       ,
       0.5       , 0.5       , 0.5       ])

In [142]: betainc(t, 3*t, 0.3)
Out[142]: 
array([0.74999936, 0.74999994, 0.74999999, 0.75      , 0.75      ,
       0.75      , 0.75      , 0.75      ])

In [143]: betainc(2*t, 3*t, 0.3)
Out[143]: 
array([0.59999898, 0.5999999 , 0.59999999, 0.6       , 0.6       ,
       0.6       , 0.6       , 0.6       ])
```

rlucas7 replied:
I think we're all in agreement on this one, consistent values here:
`a == 0, b > 0, x > 0` -> 1
`a > 0, b == 0, x > 0` ->  1
`a == 0, b == 0, x>= 0` -> `Nan`

I know Josh would like the `inf` cases handled too but not clear to me the limit value(s) for those cases so calling those  'out of scope' for this PR.

rlucas7 commented:
@person142 I restricted the fixes the both a and b finite. It's not obvious to be how the regularization interacts in the limit and whether is finite or nan.

@WarrenWeckesser I fixed the edge cases to be `a=0, b>0`, with `b` finite and `a>0, b=0` with `a` finite. 

I also added a separate branch for the `a=b=0` case which returns Nan.

rlucas7 commented:
The build failures on travis are real:
https://travis-ci.org/github/scipy/scipy/jobs/687885706#L1379

not clear to me what is the cause though, I always forget about the mpmath tests for special and can never seem to decipher the output, anyone have an idea on the fix for the failing test?

tylerjereddy commented:
For the `special` high-precision tests maybe: @person142 ?

person142 commented:
> not clear to me what is the cause though

You're hitting a real edge case-when `x = 0` you're taking a double limit in `a -> 0`/`x -> 0` or `b -> 0`/`x -> 0`. And e.g. as `a -> 0` the result goes to `1`, but as `x -> 0` the result goes to `0`.

person142 commented:
The reason we weren't hitting this before is because the test:

https://github.com/scipy/scipy/blob/master/scipy/special/tests/test_mpmath.py#L872

uses the dreaded `nan_ok=True`:

https://github.com/scipy/scipy/blob/master/scipy/special/_mptestutils.py#L183

which hides all sorts of inconvenient truths.

rlucas7 commented:
@person142 thanks for the pointers, I'll not have further time this weekend for this one. 
@tylerjereddy I will pick back up after the 1.5.0 branch. I'll punt the milestone to 1.6.0

rlucas7 commented on scipy/special/tests/test_betainc.py:
```diff
@@ -0,0 +1,29 @@
+from numpy.testing import assert_allclose 
+
+
+class Testbetainc(object):
```
right, IIRC we're no longer supporting python 2 versions, removed.

rlucas7 commented:
>You're hitting a real edge case-when x = 0 you're taking a double limit in a -> 0/x -> 0 or b -> 0/x -> 0. And >e.g. as a -> 0 the result goes to 1, but as x -> 0 the result goes to 0.

Right, so there are a couple cases here, IIUC, if `x>0, a=0, b=0` the value is `nan` on that face of the space. 

Not clear to me the statement: 

>And e.g. as a -> 0 the result goes to 1, but as x -> 0 the result goes to 0.
this suggests either a jump discontinuity ***or*** a code error not clear to me which is being referred to, 
@person142 can you clarify? 

My (working) understanding is that the code has (another) edge case here and the value the current code returns is `0` but should, in fact, be `nan` when `lim x-> 0+, a=0,b=0`, is that your understanding as well? 

> The reason we weren't hitting this before is because the test:
> 
> https://github.com/scipy/scipy/blob/master/scipy/special/tests/test_mpmath.py#L872
> 
> uses the dreaded `nan_ok=True`:
> 
> https://github.com/scipy/scipy/blob/master/scipy/special/_mptestutils.py#L183
> 
> which hides all sorts of inconvenient truths.

Yes, I see how that bit of code works now, thanks. Of my understanding above is correct (waiting for Josh to confirm or clarify) then we'll need to remove the edge case  `x=0, a=0,b=0` from this part of the tests.

tylerjereddy commented:
@rlucas7 @person142 Sounds like some progress was being made here, but also some potentially tricky things to solve still? As we approach the branching for 1.6.0 let me know if you think you can get this ready in time.

rlucas7 commented:
> @rlucas7 @person142 Sounds like some progress was being made here, but also some potentially tricky things to solve still? As we approach the branching for 1.6.0 let me know if you think you can get this ready in time.

Thanks for the ping, I removed the 1.6.0 milestone, I'll come back to this but not likely within the next week or two.

mdhaber commented on scipy/special/cephes/incbet.c:
```diff
@@ -72,15 +72,21 @@ static double incbcf(double a, double b, double x);
+	    return (NPY_NAN);
+
+    if ((aa == 0.0 && bb > 0.0 && npy_isfinite(bb)) ||
+		    (aa > 0.0 && bb == 0.0 && npy_isfinite(aa)))
```
Most of this looks right. I think this is the only problem.

```python3
from scipy import special

# To avoid mistakes when entering into Wolfram Alpha,
# make the argument orders match.
def BetaRegularized(x, a, b):
    return special.betainc(a, b, x)

# Values vere close to zero reported as 0.0
# Wolfram Alpha "(undefined)" reported as NaN
# When evaluating limits, a=0 or b=0 are replaced with x,
# and the limit is taken from the right. "-" means did not evaluate.

O = 1e-99
oo = 1e99
                                    # SciPy | Evaluated | Limit
print(BetaRegularized(O, O, O))     # 0.5   | NaN       | -
print(BetaRegularized(0.5, O, O))   # 0.5   | NaN       | -
print(BetaRegularized(1, O, O))     # 1.0   | NaN       | -
print(BetaRegularized(O, O, 1))     # 1.0   | 1         | 1 (both from the right)
print(BetaRegularized(0.5, O, 1))   # 1.0   | 1         | 1 (from the right)
print(BetaRegularized(1, O, 1))     # 1.0   | 1         | 1 (from the right)
print(BetaRegularized(O, 1, O))     # 0.0   | 0         | 0
print(BetaRegularized(0.5, 1, O))   # 0.0   | 0         | 0
print(BetaRegularized(1, 1, O))     # 1.0   | 0         | 1 (from the right)
```

So it looks like when `a > 0`, `b = 0`, and `x < 1`, the result should be 0. In this case, if `x = 1`, there is an argument to be made for making the output 1.0, but Wolfram Alpha gives a different result for the limit versus just evaluating at `b = 0`, and it gives `ComplexInfinity` in Mathematica itself, so it's not super clear.

rlucas7 replied:
Thanks @mdhaber I think the stalling point for me here was mostly lack of concensus on the correct limit values. Given the ambiguity I'm not sure what's the best way to proceed for the edge cases.

Do you have thoughts or recommendations on how to resolve the ambiguity? 

I can make an arbitrary decision if that's what is required here but would prefer not to do so.

mdhaber replied:
ISTM that the only case (that's in scope) that is not clear-cut is `a > 0`, `b == 0`, and `x == 1`. For that, I think it would be fine to continue to return NaN. If someone enjoys this sort of thing, they can write up a paper about it; if we're convinced, we can change the value then. And at some point, there will be a meta-issue about auditing edge cases like this throughout `special`; this resolves the reported issue, so we can close gh-8411 once this merges.

So if you can update this to make the changes in `incbet.h` (and maybe make the changes for `betaincc`, too), I think it would be safe enough for me to merge it.

steppi replied:
Some thoughts: Strictly speaking, if $b > 0$ then the regular pointwise limit

$$\lim_{(a, x) \rightarrow (0_+, 0_+)} \mathrm{betainc}(a, b, x)$$

should be indeterminate since the value depends on the path taken for the limit. This is also the case for $a > 0$ and

$$\lim_{(b, x) \rightarrow (0_+, 0_+)} \mathrm{betainc}(a, b, x)$$

but I sympathize: the cases $a \rightarrow 0_+$, $b > 0$ and $b \rightarrow 0_+$, $a > 0$ are less indeterminate in a sense than the case $(a, b) \rightarrow (0_+, 0_+)$, because in the former cases at least the limiting associated Beta distributions converge in distribution to a point distribution at $x = 0$ and a point distribution at $x = 1$ respectively. In the general case, the associated Beta distribution could converge to any Bernoulli distribution depending on the path taken for the limit.

I'd already suggested treating cases like this as indeterminate here, https://github.com/scipy/scipy/pull/20085#issuecomment-2028478904, in the context of defending the current behavior of `gammainc`. I think I prefer to stay consistent with `gammainc` and suggest:

```python
# betainc(a, b, x)
betainc(0, 0, 0)   # NaN
betainc(0, 0, 0.5) # NaN
betainc(0, 0, 1)   # NaN
betainc(0, 1, 0)   # NaN
betainc(0, 1, 0.5) # 1
betainc(0, 1, 1)   # 1
betainc(1, 0, 0)   # 0
betainc(1, 0, 0.5) # 0
betainc(1, 0, 1)   # NaN
```

If we want to have a separate `beta_cdf` special function, even if just private to `stats` at first, I would support

```python
beta_cdf(0, 1, +0) # 1
beta_cdf(0, 1, -0) # 0
```

a key difference between `betainc` and `beta_cdf` is for `betainc` should be extended beyond $x \in [0, 1]$ by analytic continuation and `beta_cdf` by the definition of a CDF.

mdhaber replied:
If we had a `beta_cdf`, would it be consistent with what we need for equation 3 of https://github.com/scipy/scipy/pull/11788#discussion_r1924897501 to give us the right result at the edge cases? If so, I'd like a private version of that.

steppi replied:
Actually, thinking about it more. It's not so clear cut. There's two ways that I think our reasonable for us to think about `betainc`:

1. A function of three variables `a`, `b`, `x`.
2. A two parameter family of functions of a single variable `x`.

Notations like $I_x(a, b)$ for the regularized incomplete Beta function and $B(x; a, b)$ clearly show such a split in terms of how the arguments are interpreted. The limits you suggested fall in line with interpretation 2 above, and the ones I suggested fall in line with interpretation 1. 

I actually think I might prefer sticking with interpretation 2) above, as long as the assumptions are clearly documented, and we try to be consistent within `special`. I'd be a bit embarrassed to walk  back what I said in https://github.com/jax-ml/jax/issues/20507#issue-2216680734, but it looks like no action was taken on the Jax side yet, so it's not like we'd be introducing churn if we update `gammainc` and `gammaincc` to fall in line with interpretation 2.
